### PR TITLE
Fix timezone (#895, #923)

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Solve bug #895, #923

A few test run is included

no time date selector

```python
import datetime as dt
from src.taipy.gui import Gui
from dateutil import parser

selected_date = dt.date.today()

md = """
# Timezone issue

<|{selected_date}|date|not with_time|>

<|{selected_date}|>

<|{repr(selected_date)}|>
"""


if __name__ == "__main__":
    gui = Gui(md)
    gui.run(time_zone="Etc/UTC")
```

table with no time
```python
import pandas as pd
from datetime import datetime
import src.taipy.gui.builder as tgb
from src.taipy.gui import Gui

start_date = datetime(2023, 1, 1)
end_date = datetime(2023, 2, 1)
date_range = pd.date_range(start=start_date, end=end_date)

df = pd.DataFrame({'Date': date_range})
df["Index"] = df.index

print(df)

def on_edit(state, var_name, action, payload):
    print(var_name, action, payload)

# with tgb.Page() as page:
#     tgb.table(data="{df}", editable=True, filter=True, on_edit=on_edit)

page = """
<|{df}|table|on_edit=on_edit|editable=True|filter=True|>
"""

Gui(page=page).run()
```

table with time
```python
import pandas as pd
from datetime import datetime
import src.taipy.gui.builder as tgb
from src.taipy.gui import Gui

start_datetime = datetime(2023, 1, 1, 0, 0, 0)
end_datetime = datetime(2023, 1, 3, 23, 59, 59)
datetime_range = pd.date_range(start=start_datetime, end=end_datetime, freq='H')

df = pd.DataFrame({'Date': datetime_range})
df["Index"] = df.index

print(df)

def on_edit(state, var_name, action, payload):
    print(var_name, action, payload)

# with tgb.Page() as page:
#     tgb.table(data="{df}", editable=True, filter=True, on_edit=on_edit)

page = """
<|{df}|table|on_edit=on_edit|editable=True|filter=True|date_format=yyyy-MM-dd HH:mm:ss zzz|>
"""

Gui(page=page).run()
```

### CODE REVIEW

File: gui/src/components/Taipy/DateSelector.spec.tsx [hunk 1]
```diff
@@ -143,7 +143,7 @@ describe("DateSelector Component", () => {
             await userEvent.type(input, "{ArrowLeft}{ArrowLeft}{ArrowLeft}01012001", { delay: 1 });
             expect(dispatch).toHaveBeenLastCalledWith({
                 name: "",
-                payload: { value: "2001-01-01T00:00:00.000Z" },
+                payload: { value: "Mon Jan 01 2001" },
```
Comment (FredLL-Avaiga) [inline gui/src/components/Taipy/DateSelector.spec.tsx:146]: should we not format date as YYYMMDD ?

File: src/taipy/gui/utils/date.py [hunk 2]
```diff
@@ -30,7 +31,9 @@ def _date_to_ISO(date_val: t.Union[datetime, date, time]) -> str:
     return date_val.isoformat()
 
 
-def _ISO_to_date(date_str: str) -> datetime:
+def _ISO_to_date(date_str: str) -> t.Union[datetime, date]:
     # return datetime.strptime(date_str, '%Y-%m-%dT%H:%M:%S.%fZ')
     # return datetime.fromisoformat(date_str)
-    return parser.parse(date_str)
+    date = parser.parse(date_str)
+    date_regex = r"^[A-Z][a-z]{2} (?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d{2} \d{4}$"
```
Comment (FredLL-Avaiga) [inline src/taipy/gui/utils/date.py:38]: what happens if the browser local has an impact on toDateString ?

**Review by FredLL-Avaiga — APPROVED** [review]
I think we should use a local agnostic date format like YYYMMDD

File: src/taipy/gui/utils/date.py (hunk 2 above)
Comment (dinhlongviolin1) [reply src/taipy/gui/utils/date.py:38]: it should not according to mdn as it is compatible with all browsers

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toDateString

File: gui/src/components/Taipy/DateSelector.spec.tsx (hunk 1 above)
Comment (FabienLelaquais) [reply gui/src/components/Taipy/DateSelector.spec.tsx:146]: Indeed we should be using ISO 8601 (=> "YYYYMMDD") whenever a date is serialized

Comment (dinhlongviolin1) [reply gui/src/components/Taipy/DateSelector.spec.tsx:146]: this is the default format of toDateString(). if you think it is better to change it to another format then I will do it :D

File: gui/src/utils/hooks.ts [hunk 3]
```diff
@@ -105,9 +105,10 @@ export const useFormatConfig = (): FormatConfig => {
                 timeZone: state.timeZone || TIMEZONE_CLIENT,
                 forceTZ: !!state.timeZone,
                 dateTime: state.dateTimeFormat || "yyyy-MM-dd HH:mm:ss zzz",
+                date: state.dateFormat || "yyyy/MM/dd",
```
Comment (FabienLelaquais) [inline gui/src/utils/hooks.ts]: I suspect the proper format should be "yyyy-MM-dd" (homogeneous with the datetime format)

File: gui/src/components/Taipy/DateSelector.spec.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply gui/src/components/Taipy/DateSelector.spec.tsx:146]: indeed toDateString is always in english ...
If python knows how to parse this by default, I think that we're good

File: src/taipy/gui/utils/date.py (hunk 2 above)
Comment (FabienLelaquais) [reply src/taipy/gui/utils/date.py:38]: If input is ISO (as the function name infers) then this pattern should never occur

Comment (dinhlongviolin1) [reply src/taipy/gui/utils/date.py:38]: Oh, do you mean that the browser timezone would affect the date parsing

File: tests/taipy/gui/e2e/test_timezone.py [hunk 4]
```diff
@@ -65,3 +65,18 @@ def _timezone_test_template(page: "Page", gui: Gui, helpers, time_zone, texts):
     page.wait_for_selector("#text1")
     text1 = page.query_selector("#text1")
     assert text1.inner_text() in texts
+
+
+def test_date_only(page: "Page", gui: Gui, helpers):
+    page_md = """
+<|{t}|id=text1|>
+"""
+    t = _ISO_to_date("Wed Jul 28 1993")  # noqa: F841
```
Comment (FabienLelaquais) [inline tests/taipy/gui/e2e/test_timezone.py]: That is not ISO... or the function is wrongly named, and ENG-locale dependent... which is bad

File: gui/src/components/Taipy/DateSelector.spec.tsx (hunk 1 above)
Comment (dinhlongviolin1) [reply gui/src/components/Taipy/DateSelector.spec.tsx:146]: It knows pretty well. Tested it and it works like a charm

File: tests/taipy/gui/e2e/test_timezone.py [hunk 5]
```diff
@@ -65,3 +65,18 @@ def _timezone_test_template(page: "Page", gui: Gui, helpers, time_zone, texts):
...
+    gui._set_frame(inspect.currentframe())
+    gui.add_page(name="test", page=page_md)
+    helpers.run_e2e(gui)
+    page.goto("./test")
+    page.expect_websocket()
+    page.wait_for_selector("#text1")
+    text1 = page.query_selector("#text1")
+    assert text1.inner_text() in ["1993/07/28"]
```
Comment (FabienLelaquais) [inline tests/taipy/gui/e2e/test_timezone.py]: 1993-07-28 (?)

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Make sure ISO is used all over the place - we can discuss that anytime.

File: src/taipy/gui/utils/date.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/utils/date.py:38]: it could
and if you use YYY-MM-DD you can simply test the length

**Review by FabienLelaquais — APPROVED** [review]
👌

File: src/taipy/gui/utils/date.py (hunk 2 above)
Comment (dinhlongviolin1) [reply src/taipy/gui/utils/date.py:38]: right now, the toDateString method is the one being used when withTime = False so we dont have to concern with other cases such as YYYY-mm-dd. For the affect of tz on the date parsing, I trust that the date selector would do the job of getting the correct date :D

**Pull request merged**